File: common/src/main/java/cc/turtl/chiselmon/system/tracker/TrackerSession.java
package cc.turtl.chiselmon.system.tracker;

import cc.turtl.chiselmon.ChiselmonStorage;
import cc.turtl.chiselmon.client.api.PokemonLoadedEvent;
import cc.turtl.chiselmon.client.api.PokemonUnloadedEvent;
import cc.turtl.chiselmon.api.storage.StorageScope;
import cc.turtl.chiselmon.core.api.PokemonEncounter;
import cc.turtl.chiselmon.system.dashboard.DashboardServer;
import cc.turtl.chiselmon.util.render.PokemonEntityUtils;
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity;

import java.io.IOException;
import java.util.*;

/**
 * Keeps track of currently loaded WILD pokemon, and stores all new encounters in a database.
 */
public class TrackerSession {
    private static final int DASHBOARD_PORT = 7890;

    private final long startTimeMs;
    private final Map<UUID, PokemonEntity> currentlyLoaded = new HashMap<>();
    private final Set<UUID> seenUuids = new HashSet<>();
    private final EncounterDatabase db;
    private DashboardServer dashboardServer;

    public TrackerSession() {
        this.startTimeMs = System.currentTimeMillis();
        this.db = ChiselmonStorage.ENCOUNTERS.get(StorageScope.currentWorld());
    }

    public void onPokemonLoad(PokemonLoadedEvent event) {
        if (!event.isWild()) return;

        PokemonEncounter encounter = event.getEncounter();
        UUID uuid = event.getEntity().getUUID();

        currentlyLoaded.put(uuid, event.getEntity());
        seenUuids.add(uuid);
        db.record(encounter);
    }

    public void onPokemonUnload(PokemonUnloadedEvent event) {
        currentlyLoaded.remove(event.getEntity().getUUID());
    }

    public void tick() {
        cleanUnloaded();

        // reset these always, before despawn glow or alert, so they always work from a blank slate.
        for (PokemonEntity entity : currentlyLoaded.values()) {
            PokemonEntityUtils.removeGlow(entity);
            PokemonEntityUtils.resetNickname(entity);
        }
    }

    private void cleanUnloaded() {
        currentlyLoaded.entrySet().removeIf(e -> e.getValue().isRemoved());
    }

    public EncounterDatabase getDb() {
        return db;
    }

    public void startDashboard() throws IOException {
        if (dashboardServer != null) return;
        dashboardServer = new DashboardServer(db, DASHBOARD_PORT);
        dashboardServer.start();
    }

    public void stopDashboard() {
        if (dashboardServer == null) return;
        dashboardServer.stop();
        dashboardServer = null;
    }

    public boolean isDashboardRunning() {
        return dashboardServer != null;
    }

    public long dashboardUptime() {
        return dashboardServer.uptime();
    }

    public int getDashboardPort() {
        return DASHBOARD_PORT;
    }

    public Map<UUID, PokemonEntity> getCurrentlyLoaded() {
        return currentlyLoaded;
    }

    public long getMsElapsed() {
        return System.currentTimeMillis() - startTimeMs;
    }

    public int getEncounterCount() {
        return seenUuids.size();
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/system/tracker/TrackerSession.kt
package cc.turtl.chiselmon.system.tracker

import cc.turtl.chiselmon.client.ChiselmonStorage
import cc.turtl.chiselmon.client.api.PokemonLoadedEvent
import cc.turtl.chiselmon.client.api.PokemonUnloadedEvent
import cc.turtl.chiselmon.client.util.removeGlow
import cc.turtl.chiselmon.client.util.resetNickname
import cc.turtl.chiselmon.core.api.storage.Scope
import cc.turtl.chiselmon.system.dashboard.DashboardServer
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity
import java.util.UUID

/**
 * Keeps track of currently loaded WILD pokemon, and stores all new encounters in a database.
 */
class TrackerSession {

    private val startTimeMs: Long = System.currentTimeMillis()
    private val seenUuids: MutableSet<UUID> = HashSet()

    val currentlyLoaded: MutableMap<UUID, PokemonEntity> = HashMap()
    val db: EncounterDatabase = run {
        val worldScope = Scope.currentWorld()
            ?: error("TrackerSession must be created while in a world")
        ChiselmonStorage.ENCOUNTERS[worldScope]
    }

    private var dashboardServer: DashboardServer? = null

    fun onPokemonLoad(event: PokemonLoadedEvent) {
        if (!event.isWild) return

        val uuid = event.entity.uuid
        currentlyLoaded[uuid] = event.entity
        seenUuids.add(uuid)
        db.record(event.encounter)
    }

    fun onPokemonUnload(event: PokemonUnloadedEvent) {
        currentlyLoaded.remove(event.entity.uuid)
    }

    fun tick() {
        cleanUnloaded()

        // reset these always, before despawn glow or alert, so they always work from a blank slate.
        for (entity in currentlyLoaded.values) {
            entity.removeGlow()
            entity.resetNickname()
        }
    }

    private fun cleanUnloaded() {
        currentlyLoaded.entries.removeAll { it.value.isRemoved }
    }

    fun startDashboard() {
        if (dashboardServer != null) return
        dashboardServer = DashboardServer(db, DASHBOARD_PORT).also { it.start() }
    }

    fun stopDashboard() {
        dashboardServer?.stop()
        dashboardServer = null
    }

    val isDashboardRunning: Boolean
        get() = dashboardServer != null

    fun dashboardUptime(): Long = dashboardServer?.uptime() ?: 0L

    val dashboardPort: Int get() = DASHBOARD_PORT

    val msElapsed: Long
        get() = System.currentTimeMillis() - startTimeMs

    val encounterCount: Int
        get() = seenUuids.size

    companion object {
        private const val DASHBOARD_PORT = 7890
    }
}
